File: svws-core/src/main/java/de/svws_nrw/core/abschluss/bk/d/BKGymBelegpruefungD3.java
package de.svws_nrw.core.abschluss.bk.d;

import java.util.List;
import java.util.Map;

import de.svws_nrw.asd.data.schule.BeruflichesGymnasiumStundentafel;
import de.svws_nrw.asd.data.schule.BeruflichesGymnasiumStundentafelAbiturfaecherWahlmoeglichkeit;
import de.svws_nrw.asd.types.schule.BeruflichesGymnasiumPruefungsordnungAnlage;
import jakarta.validation.constraints.NotNull;

/**
 * Der Belegprüfungsalgorithmus für den Bildungsgang der Schulgliederung D01
 * und der Fachklasse 106 00.
 */
public final class BKGymBelegpruefungD3 extends BKGymBelegpruefung {

	/**
	 * Erzeugt einen neue Belegprüfung
	 *
	 * @param manager   der Manager für die Abiturdaten
	 */
	public BKGymBelegpruefungD3(@NotNull final BKGymAbiturdatenManager manager) {
		super(manager);
	}

	@Override
	public void pruefe() {
		// Prüfe die Abiturfächer darauf, ob diese bei mehreren Fachbelegungen eingetragen sind

		// Bestimme zunächst die möglichen Stundentafeln laut der Anlage und der beiden belegten Leistungskurse.
		final @NotNull List<BeruflichesGymnasiumStundentafel> mglStundentafeln = getStundentafelnByAbiturfaechern(BeruflichesGymnasiumPruefungsordnungAnlage.D3);

		// Verlasse die Prüfung, wenn bis hier keine Variante gefunden wurde
		if (mglStundentafeln.isEmpty())
			return;

		// Prüfe die Wahlmöglichkeiten bei den Varianten in Bezug auf AB3 und AB4, verlasse ggf. die Prüfung, wenn keine Wahlmöglichkeit für AB3 und AB4 gefunden wurde
		final @NotNull Map<BeruflichesGymnasiumStundentafel, BeruflichesGymnasiumStundentafelAbiturfaecherWahlmoeglichkeit> mapWahlmoeglichkeiten = manager.getWahlmoeglichekeiten(mglStundentafeln);
		if (mapWahlmoeglichkeiten.isEmpty()) {
			addFehler(BKGymBelegungsfehler.AB_5);
			return;
		}


		// 1. Abiturfachkombination prüfen und Zuordnung der Abiturfachs zum Fach (O punkte als nicht belegt)

		// 2. Schritt identifikation der Variante:

		// 3. restliche Fachbelegung prüfen
	}

}

File: svws-webclient/core/src/core/abschluss/bk/d/BKGymBelegpruefungD3.ts
import { BeruflichesGymnasiumStundentafelAbiturfaecherWahlmoeglichkeit } from '../../../../asd/data/schule/BeruflichesGymnasiumStundentafelAbiturfaecherWahlmoeglichkeit';
import { BKGymBelegungsfehler } from '../../../../core/abschluss/bk/d/BKGymBelegungsfehler';
import { BKGymAbiturdatenManager } from '../../../../core/abschluss/bk/d/BKGymAbiturdatenManager';
import { BKGymBelegpruefung, cast_de_svws_nrw_core_abschluss_bk_d_BKGymBelegpruefung } from '../../../../core/abschluss/bk/d/BKGymBelegpruefung';
import { BeruflichesGymnasiumPruefungsordnungAnlage } from '../../../../asd/types/schule/BeruflichesGymnasiumPruefungsordnungAnlage';
import type { List } from '../../../../java/util/List';
import { BeruflichesGymnasiumStundentafel } from '../../../../asd/data/schule/BeruflichesGymnasiumStundentafel';
import { Class } from '../../../../java/lang/Class';
import type { JavaMap } from '../../../../java/util/JavaMap';

export class BKGymBelegpruefungD3 extends BKGymBelegpruefung {


	/**
	 * Erzeugt einen neue Belegprüfung
	 *
	 * @param manager   der Manager für die Abiturdaten
	 */
	public constructor(manager: BKGymAbiturdatenManager) {
		super(manager);
	}

	public pruefe(): void {
		const mglStundentafeln: List<BeruflichesGymnasiumStundentafel> = this.getStundentafelnByAbiturfaechern(BeruflichesGymnasiumPruefungsordnungAnlage.D3);
		if (mglStundentafeln.isEmpty())
			return;
		const mapWahlmoeglichkeiten: JavaMap<BeruflichesGymnasiumStundentafel, BeruflichesGymnasiumStundentafelAbiturfaecherWahlmoeglichkeit> = this.manager.getWahlmoeglichekeiten(mglStundentafeln);
		if (mapWahlmoeglichkeiten.isEmpty()) {
			this.addFehler(BKGymBelegungsfehler.AB_5);
			return;
		}
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.core.abschluss.bk.d.BKGymBelegpruefungD3';
	}

	isTranspiledInstanceOf(name: string): boolean {
		return ['de.svws_nrw.core.abschluss.bk.d.BKGymBelegpruefungD3', 'de.svws_nrw.core.abschluss.bk.d.BKGymBelegpruefung'].includes(name);
	}

	public static class = new Class<BKGymBelegpruefungD3>('de.svws_nrw.core.abschluss.bk.d.BKGymBelegpruefungD3');

}

export function cast_de_svws_nrw_core_abschluss_bk_d_BKGymBelegpruefungD3(obj: unknown): BKGymBelegpruefungD3 {
	return obj as BKGymBelegpruefungD3;
}
